File: android/src/main/java/com/urbanairship/reactnative/ReactMessageViewManager.java
/* Copyright Airship and Contributors */

package com.urbanairship.reactnative;


import androidx.annotation.Nullable;
import androidx.annotation.NonNull;

import com.facebook.react.common.MapBuilder;
import com.facebook.react.uimanager.SimpleViewManager;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.annotations.ReactProp;

import java.util.Arrays;
import java.util.List;
import java.util.Map;

public class ReactMessageViewManager extends SimpleViewManager<ReactMessageView> {

    public static final String REACT_CLASS = "UARCTMessageView";

    @Override
    public String getName() {
        return REACT_CLASS;
    }

    @Nonnull
    @Override
    protected ReactMessageView createViewInstance(@Nonnull ThemedReactContext reactContext) {
        return new ReactMessageView(reactContext);
    }

    @ReactProp(name = "messageId")
    public void setMessageId(ReactMessageView view, @Nullable String messageId) {
        view.loadMessage(messageId);
    }

    public Map getExportedCustomBubblingEventTypeConstants() {
        List<String> events = Arrays.asList(ReactMessageView.EVENT_CLOSE,
                ReactMessageView.EVENT_LOAD_ERROR,
                ReactMessageView.EVENT_LOAD_FINISHED,
                ReactMessageView.EVENT_LOAD_STARTED);

        MapBuilder.Builder builder = MapBuilder.builder();

        for (String event : events) {
            builder.put(event,
                    MapBuilder.of(
                    "phasedRegistrationNames",
                    MapBuilder.of("bubbled", event)));
        }

        return builder.build();
    }


}

File: android/src/main/java/com/urbanairship/reactnative/ReactMessageViewManager.kt
/* Copyright Airship and Contributors */

package com.urbanairship.reactnative

import com.facebook.react.common.MapBuilder
import com.facebook.react.uimanager.SimpleViewManager
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.ViewManagerDelegate
import com.facebook.react.uimanager.annotations.ReactProp
import com.facebook.react.viewmanagers.RNAirshipMessageViewManagerInterface
import com.facebook.react.bridge.ReadableArray

class ReactMessageViewManager : SimpleViewManager<ReactMessageView>(),
    RNAirshipMessageViewManagerInterface<ReactMessageView> {

    private val delegate = object : ViewManagerDelegate<ReactMessageView> {

        override fun setProperty(view: ReactMessageView, propName: String, value: Any?) {
            when (propName) {
                "messageId" -> setMessageId(view, value as? String)
                else -> {}
            }
        }

        override fun receiveCommand(
            view: ReactMessageView,
            commandName: String,
            args: ReadableArray
        ) {
            // No commands supported
        }
    }

    override fun getName(): String {
        return REACT_CLASS
    }

    override fun getDelegate(): ViewManagerDelegate<ReactMessageView> {
        return delegate
    }

    override fun createViewInstance(reactContext: ThemedReactContext): ReactMessageView {
        val messageView = ReactMessageView(reactContext)
        reactContext.addLifecycleEventListener(messageView)
        return messageView
    }

    override fun onDropViewInstance(messageView: ReactMessageView) {
        super.onDropViewInstance(messageView)
        (messageView.context as ThemedReactContext).removeLifecycleEventListener(messageView)
        messageView.cleanup()
    }

    @ReactProp(name = "messageId")
    override fun setMessageId(view: ReactMessageView, messageId: String?) {
        messageId?.let {
            view.loadMessage(it)
        }
    }

    override fun getExportedCustomBubblingEventTypeConstants(): Map<String, Any> {
        val events = listOf(
                ReactMessageView.EVENT_CLOSE_REGISTRATION_NAME to ReactMessageView.EVENT_CLOSE_HANDLER_NAME,
                ReactMessageView.EVENT_LOAD_ERROR_REGISTRATION_NAME to ReactMessageView.EVENT_LOAD_ERROR_HANDLER_NAME,
                ReactMessageView.EVENT_LOAD_FINISHED_REGISTRATION_NAME to ReactMessageView.EVENT_LOAD_FINISHED_HANDLER_NAME,
                ReactMessageView.EVENT_LOAD_STARTED_REGISTRATION_NAME to ReactMessageView.EVENT_LOAD_STARTED_HANDLER_NAME
            )

        val builder = MapBuilder.builder<String, Any>()

        for ((name, handlerName) in events) {
            builder.put(
                name,
                MapBuilder.of(
                    "phasedRegistrationNames",
                    MapBuilder.of("bubbled", handlerName)
                )
            )
        }

        return builder.build()
    }

    companion object {
        const val REACT_CLASS = "RNAirshipMessageView"
    }
}
